lib/Parse/ParseType.cpp

```diff
@@ -817,6 +827,79 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclBaseType) {
   return makeParserResult(Status, DeclRefTR);
 }
 
+ParserResult<TypeRepr> Parser::parseTypeIdentifier() {
+  // FIXME: We should parse e.g. 'X.var'. Almost any keyword is a valid member
+  // component.
+  DeclNameLoc Loc;
+  DeclNameRef Name =
+      parseDeclNameRef(Loc, diag::expected_identifier_in_dotted_type, {});
+  if (!Name)
+    return makeParserError();
+
+  ParserStatus Status;
+  IdentTypeRepr *IdTR;
+
+  if (startsWithLess(Tok)) {
+    SourceLoc LAngle, RAngle;
+    SmallVector<TypeRepr *, 8> GenericArgs;
+    auto ArgsStatus = parseGenericArguments(GenericArgs, LAngle, RAngle);
+    if (ArgsStatus.isErrorOrHasCompletion())
+      return ArgsStatus;
+
+    IdTR = GenericIdentTypeRepr::create(Context, Loc, Name, GenericArgs,
+                                        SourceRange(LAngle, RAngle));
+  } else {
+    IdTR = new (Context) SimpleIdentTypeRepr(Loc, Name);
+  }
+
+  return makeParserResult(IdTR);
+}
+
+ParserResult<TypeRepr> Parser::parseTypeDotted(ParserResult<TypeRepr> Base) {
+  assert(Base.isNonNull());
+  assert(Tok.isAny(tok::period, tok::period_prefix));
+
+  SmallVector<IdentTypeRepr *, 4> MemberComponents;
+
+  while (true) {
+    if (Tok.isNot(tok::period, tok::period_prefix) ||
+        peekToken().is(tok::code_complete)) {
+      break;
+    }
+
+    // Consume the period.
+    consumeToken();
+
+    if (Tok.isContextualKeyword("Type") ||
+        Tok.isContextualKeyword("Protocol")) {
+      TypeRepr *MetaBase =
+          MemberTypeRepr::create(Context, Base.get(), MemberComponents);
```

Comment:
Oh I see. `MemberTypeRepr::create()` just returns `Base` when the components is empty. hmm...

Reply:
> Oh I see. `MemberTypeRepr::create()` just returns `Base` when the components is empty. hmm...

Perhaps moving this factory constructor to `DeclRefTypeRepr` is a better idea. We had 4 or 5 places that had to do the if-else on the components’ size just to make the constructor happy, so I revised the separation of concerns for convenience (see the first commit).

Reply:
Yeah, `DeclRefTypeRepr::create()` sounds more natural.

Reply:
@rintaro I‘ve been turning this over in my mind and I believe it would only complicate things. We have to keep `MemberTypeRepr::create` either way, meaning that we’d have 2 factory methods with identical signatures and a subtle difference in contract spread across different types. It also seems more intuitive for a factory method that returns `X` or `Y` to be declared on the meet, i.e., the more specific type.

Reply:
Agreed.